<a href="https://colab.research.google.com/github/Nathan-oc/Estudos-DSNP/blob/main/Introdu%C3%A7%C3%A3o_ao_Pandas_INSPIRADO.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Introdução ao Pandas**

  `Pandas` é a principal bilbioteca do Python quando estamos falando de Data Science. 
  
  Com ela é possível importar base de dados (arquivos `csv` e `xlsx`, por exemplo), tratar/limpar esses dados, transformá-los e realizar análises completas dos mesmos. 

  Uma vez que importamos um conjunto de dados utilizando o Pandas, fica muito simples fazer coisas do tipo:
  * Extrair informações estatísticas como:
    * Qual é a média, mediana, valores máximos e mínimos?
    * Qual é a distribuição das variáveis?
    * Qual é a correlação entre duas variáveis quaisquer?
  * Exportar os dados para um novo formato de arquivo.
  * Visualizar gráficos dos mais diferentes tipos.
  * Alimentar modelos de *machine learning* feitos em cima de `scikit-learn` ou alguma outra biblioteca.


O Pandas é construído em cima de outra biblioteca muito famosa, a bibliteca `Numpy`, devido a isso, aquela possui grandes similaridades com essa. 

---
## Importando a biblioteca

Para podermos utilizar precisamos instalar a bibioteca no computador, caso o uso seja em alguma IDE da máquina local. Porém, como estamos utilizando *Google Colab*, a bilblioteca já está instalado e precisamos apenas importar para esse notebook.

In [ ]:
import pandas as pd
import numpy as np

---
## Obtendo dados

* ### Importação a partir de arquivos

Para dar um exemplo de como podemos facilmente importar arquivos, baixei uma planilha em .csv do preço diário da ação da Petrobrás, *PETR3.SA*, no site [Yahoo Finance](https://finance.yahoo.com/quote/PETR3.SA/history?p=PETR3.SA) e disponibilizei [neste link](https://raw.githubusercontent.com/Nathan-oc/Estudos-DSNP/main/M%C3%B3dulo%202/PETR3.SA.csv?token=GHSAT0AAAAAAB4Q3V22HSOLHCWE4AQRIG6YY5HJTSQ).

Agora vamos importar facilmente essa base de dados para nosso notebook a partir do link na web (pode ser o endereço na máquina local também).

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Nathan-oc/Estudos-DSNP/main/M%C3%B3dulo%202/PETR3.SA.csv")

Com isso temos a base de dados importada para o nosso notebook.

* ### Criação a partir de variáveis já existentes

Podemos também converter um objeto do Python como uma lista ou dicionário em um dataframe a partir da função `pd.DataFrame`

* ### Salvar arquivos

Além disso, podemos salvar o arquivo para algum local com a função `pd.to_filetype(filename)`

In [ ]:
df.to_csv("PETR3.csv")

---
## Componentes básicos do Pandas

Os dois principais componentes do Pandas são  `Dataframe` e `Series`. Simplificadamente podemos pensar o primeiro como a planilha inteira dos dados, enquanto o segundo é apenas uma coluna dessa planilha.  

Essas duas estrututuras apesar de serem simples, elas facilitam muito o trabalho com dados, uma vez que podem armazenar qualquer tipo de dados (como int, float, string, data, lista etc).


In [ ]:
# ver o tipo da variável df
type(df)

pandas.core.frame.DataFrame

In [ ]:
# ver o tipo de uma coluna do dataframe
type(df['Date'])

pandas.core.series.Series

## Conhecendo os dados e funções do Pandas

* ###  Visualização rápida


Vamos dar uma olhada nos dados com `df.head()` e `df.tail()` pois é muito importante termos noção qual é a "cara" dos dados que estamos lidando:

In [ ]:
df.head()

Date       Open       High        Low      Close  Adj Close    Volume
0  2021-12-27  30.440001  31.230000  30.240000  31.040001  19.029440  17868100
1  2021-12-28  31.260000  31.469999  31.000000  31.059999  19.041702   8078000
2  2021-12-29  31.010000  31.340000  30.750000  30.950001  18.974266  11460950
3  2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001  34157000
4  2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711  15578500

In [ ]:
df.tail()

Date       Open       High        Low      Close  Adj Close  \
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001  25.910000  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999  26.680000  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000  27.320000  28.559999  28.559999   

       Volume  
245  12963900  
246  15549900  
247  20946900  
248  14139200  
249  12490700

* ### Informações do dataframe

Vamos agora extrair algumas informações subjacentes do DataFrame a partir das seguintes funções:


* `df.shape`: retorna o número de linhas e colunas
* `df.index`: retorna as labels das linhas
* `df.columns`: retorna os nomes das colunas
* `df.dtypes`: retorna os tipos de cada coluna
* `df.info()`: retorna informações de tipos de objeto, valores não ausentes e uso da memória a respeito do dataframe
* `df.select_dtypes(include=None, exclude=None)`: retorna o dataframe apenas com os tipos de dados selecionados (int, float, data, string, Bool etc)
* `df.values`: retorna um array dos valores do dataframe
* `df.axes`: retorna os nomes/labels/rótulos das colunas e linhas 
* `df.ndim`: retorna o número dimensões do data frame (geralmente é 2 mas por exemplo uma série possui uma dimensão
* `df.size`: retorna o número de elementos do dataframe, geralmente é o produto do número de linhas pelo número de colunas (se não houver valores ausentes)
* `df.empty`: retorna um booleano indicando se é um dataframe vazio ou não

In [ ]:
df.shape

(250, 7)

In [ ]:
df.index

RangeIndex(start=0, stop=250, step=1)

In [ ]:
df.columns

Index(['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume'], dtype='object')

In [ ]:
df.dtypes

Date          object
Open         float64
High         float64
Low          float64
Close        float64
Adj Close    float64
Volume         int64
dtype: object

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       250 non-null    object 
 1   Open       250 non-null    float64
 2   High       250 non-null    float64
 3   Low        250 non-null    float64
 4   Close      250 non-null    float64
 5   Adj Close  250 non-null    float64
 6   Volume     250 non-null    int64  
dtypes: float64(5), int64(1), object(1)
memory usage: 13.8+ KB


In [ ]:
df.select_dtypes(exclude=int)

Date       Open       High        Low      Close  Adj Close
0    2021-12-27  30.440001  31.230000  30.240000  31.040001  19.029440
1    2021-12-28  31.260000  31.469999  31.000000  31.059999  19.041702
2    2021-12-29  31.010000  31.340000  30.750000  30.950001  18.974266
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001
4    2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711
..          ...        ...        ...        ...        ...        ...
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999
246  2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000
247  2022-12-21  26.559999  26.860001  25.910000  26.860001  26.860001
248  2022-12-22  27.100000  27.719999  26.680000  27.170000  27.170000
249  2022-12-23  27.389999  28.600000  27.320000  28.559999  28.559999

[250 rows x 6 columns]

In [ ]:
df.values

array([['2021-12-27', 30.440001, 31.23, ..., 31.040001, 19.02944,
        17868100],
       ['2021-12-28', 31.26, 31.469999, ..., 31.059999, 19.041702,
        8078000],
       ['2021-12-29', 31.01, 31.34, ..., 30.950001, 18.974266, 11460950],
       ...,
       ['2022-12-21', 26.559999, 26.860001, ..., 26.860001, 26.860001,
        20946900],
       ['2022-12-22', 27.1, 27.719999, ..., 27.17, 27.17, 14139200],
       ['2022-12-23', 27.389999, 28.6, ..., 28.559999, 28.559999,
        12490700]], dtype=object)

In [ ]:
df.axes

[RangeIndex(start=0, stop=250, step=1),
 Index(['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume'], dtype='object')]

In [ ]:
df.ndim

2

In [ ]:
df.High.ndim

1

In [ ]:
df.size

1750

In [ ]:
df.empty

False

Uma função importante no Pandas é `df.copy()` pois podemos fazer uma cópia do dataframe original e trabalharmos em cima da cópia para que não alteremos os dados originais.

In [ ]:
df_copia = df.copy()
df_copia.head(3)

Date       Open       High    Low      Close  Adj Close    Volume
0  2021-12-27  30.440001  31.230000  30.24  31.040001  19.029440  17868100
1  2021-12-28  31.260000  31.469999  31.00  31.059999  19.041702   8078000
2  2021-12-29  31.010000  31.340000  30.75  30.950001  18.974266  11460950

* ### Estatísticas descritivas dos dados

A principal parte da análise exploratória de dados é, como o próprio nome diz, a exploração dos dados. E nada melhor para isso do que as estatísticas descritivas, que são aquelas que nos contam como os dados estão distríbuidos e suas feições.

Nessa seção iremos destrinchar diversas funções do Pandas que serão de grande utilidade:


* `df.describe()`: retorna diversas informações estatísticas de cada coluna do dataframe como número de valores, média, desvio padrão, quartis e valores mínimos e máximos.
* `df.count()`: retorna o número de valores não ausentes de cada coluna
* `df.mean()`: retorna a média de cada coluna
* `df.std()`: retorna o desvio padrão de cada coluna
* `df.min()`: retorna o valor mínimo de cada coluna
* `df.max()`: retorna o valor máximo de cada coluna
* `df.median()`: retorna a mediana de cada coluna
* `df.mode()`: retorna a moda de cada coluna

In [ ]:
df.describe()

Open        High         Low       Close   Adj Close  \
count  250.000000  250.000000  250.000000  250.000000  250.000000   
mean    33.771840   34.368800   33.176960   33.793360   25.496893   
std      3.162086    3.160505    3.119615    3.144832    4.320477   
min     23.799999   25.049999   23.610001   24.290001   18.747433   
25%     31.469999   31.910000   30.959999   31.527500   22.016315   
50%     34.370001   34.970000   33.784999   34.430000   24.288059   
75%     36.107499   36.797500   35.355001   36.235002   29.050100   
max     40.430000   42.080002   40.130001   41.560001   36.963558   

             Volume  
count  2.500000e+02  
mean   1.969298e+07  
std    9.174046e+06  
min    4.878800e+06  
25%    1.342418e+07  
50%    1.798265e+07  
75%    2.447918e+07  
max    6.977550e+07

In [ ]:
df.select_dtypes(include = object).describe()

Date
count          250
unique         250
top     2021-12-27
freq             1

In [ ]:
df.count()

Date         250
Open         250
High         250
Low          250
Close        250
Adj Close    250
Volume       250
dtype: int64

In [ ]:
df.mean()

<ipython-input-287-c61f0c8f89b5>:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df.mean()


Open         3.377184e+01
High         3.436880e+01
Low          3.317696e+01
Close        3.379336e+01
Adj Close    2.549689e+01
Volume       1.969298e+07
dtype: float64

In [ ]:
df.std()

<ipython-input-288-ce97bb7eaef8>:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df.std()


Open         3.162086e+00
High         3.160505e+00
Low          3.119615e+00
Close        3.144832e+00
Adj Close    4.320477e+00
Volume       9.174046e+06
dtype: float64

In [ ]:
df.min()

Date         2021-12-27
Open          23.799999
High          25.049999
Low           23.610001
Close         24.290001
Adj Close     18.747433
Volume          4878800
dtype: object

In [ ]:
df.max()

Date         2022-12-23
Open              40.43
High          42.080002
Low           40.130001
Close         41.560001
Adj Close     36.963558
Volume         69775500
dtype: object

In [ ]:
df.select_dtypes(exclude=[object]).median()

Open         3.437000e+01
High         3.497000e+01
Low          3.378500e+01
Close        3.443000e+01
Adj Close    2.428806e+01
Volume       1.798265e+07
dtype: float64

Aqui excluímos o tipo *object* para o Pandas lidar apenas com os valores numéricos. 

In [ ]:
df.select_dtypes(exclude=object).mode(axis='index')

Open       High        Low      Close  Adj Close    Volume
0    36.0  35.000000  32.740002  30.700001  21.438774   4878800
1     NaN  35.150002        NaN  34.689999  21.457165   5781000
2     NaN        NaN        NaN        NaN  22.235756   5827800
3     NaN        NaN        NaN        NaN  24.823196   6224000
4     NaN        NaN        NaN        NaN  27.242392   6915000
..    ...        ...        ...        ...        ...       ...
245   NaN        NaN        NaN        NaN        NaN  48903800
246   NaN        NaN        NaN        NaN        NaN  49607100
247   NaN        NaN        NaN        NaN        NaN  52964400
248   NaN        NaN        NaN        NaN        NaN  55281200
249   NaN        NaN        NaN        NaN        NaN  69775500

[250 rows x 6 columns]

Na coluna *Volume* não há valores repetidos, portanto, todos são moda. E dessa forma, o Panda mostra a planilha com todas as modas de todas as colunas.

Outras funções:

* `df.corr()`: retorna a correlação entre todas variáveis dois-a-dois
* `df.cov()`: retorna a matriz de covariância entre as variáveis
* `df.clip(lower=None, upper=None)`: limita os valores com limitante inferior e/ou superior, substituindo pelos limitantes os valores que estiverem fora do intervalo
* `df.diff(periods=0, axis=0)`: retorna um df com a diferença entre as valores de uma linha e da linha anterior (padrão)
* `df.eval(exp, inplace=False)`: recebe uma expressão em string para fazer uma operação com as colunas
* `df.pct_change(periods=1)`: retorna um df com a mudança percentual de uma linha para a linha anterior
* `df.prod()`: retorna o produto de uma coluna
* `df.quantile(q=0.5, interpolation='linear')`: retorna o valor do quantil q especificado, a partir do método de interpolação (se necessário)
* `df.rank()`: ranqueia os valores das colunas
* `df.round()`: arredonda os valores com o número de casas decimais especificadas
* `df.var()`: retorna a variância de cada coluna
* `df.nunique()`: retorna o número de valores distintos de cada coluna
* `df.value_counts()`: retorna o valor de cada entrada, o número de aparições e ordena por ordem descendente de frequência



In [ ]:
df.corr()

Open      High       Low     Close  Adj Close    Volume
Open       1.000000  0.988623  0.984565  0.969610   0.242420  0.021092
High       0.988623  1.000000  0.986291  0.987647   0.264558  0.051123
Low        0.984565  0.986291  1.000000  0.988828   0.242683 -0.039759
Close      0.969610  0.987647  0.988828  1.000000   0.258964  0.008493
Adj Close  0.242420  0.264558  0.242683  0.258964   1.000000  0.214571
Volume     0.021092  0.051123 -0.039759  0.008493   0.214571  1.000000

In [ ]:
df.cov()

Open          High           Low          Close  \
Open            9.998785  9.880085e+00  9.712232e+00       9.642023   
High            9.880085  9.988790e+00  9.724390e+00       9.816473   
Low             9.712232  9.724390e+00  9.731999e+00       9.701062   
Close           9.642023  9.816473e+00  9.701062e+00       9.889970   
Adj Close       3.311880  3.612510e+00  3.270935e+00       3.518585   
Volume     611851.282095  1.482278e+06 -1.137880e+06  245017.780567   

              Adj Close        Volume  
Open       3.311880e+00  6.118513e+05  
High       3.612510e+00  1.482278e+06  
Low        3.270935e+00 -1.137880e+06  
Close      3.518585e+00  2.450178e+05  
Adj Close  1.866652e+01  8.504794e+06  
Volume     8.504794e+06  8.416312e+13

In [ ]:
df.Volume.clip(50000000, 60000000)

0      50000000
1      50000000
2      50000000
3      50000000
4      50000000
         ...   
245    50000000
246    50000000
247    50000000
248    50000000
249    50000000
Name: Volume, Length: 250, dtype: int64

In [ ]:
df.select_dtypes(exclude=object).diff()

Open      High       Low     Close  Adj Close      Volume
0         NaN       NaN       NaN       NaN        NaN         NaN
1    0.819999  0.239999  0.760000  0.019998   0.012262  -9790100.0
2   -0.250000 -0.129999 -0.250000 -0.109998  -0.067436   3382950.0
3   -0.059999 -0.210001 -0.200001 -0.250000  -0.153265  22696050.0
4   -0.110001  0.540001  0.210001  0.819999   0.502710 -18578500.0
..        ...       ...       ...       ...        ...         ...
245  0.039999  0.590000  0.370001  0.569999   0.569999 -15159700.0
246  0.630001  0.700001  0.729999  0.620001   0.620001   2586000.0
247  1.029998  0.520001  0.490000  0.690001   0.690001   5397000.0
248  0.540001  0.859998  0.770000  0.309999   0.309999  -6807700.0
249  0.289999  0.880001  0.640000  1.389999   1.389999  -1648500.0

[250 rows x 6 columns]

In [ ]:
df.eval('Diferença = High - Low')

Date       Open       High        Low      Close  Adj Close  \
0    2021-12-27  30.440001  31.230000  30.240000  31.040001  19.029440   
1    2021-12-28  31.260000  31.469999  31.000000  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000  30.750000  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001  25.910000  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999  26.680000  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000  27.320000  28.559999  28.559999   

       Volume  Diferença  
0    17868100   0.990000  
1     8078000   0.469999  
2    11460950   0.590000  
3    34157000   0.580000  
4    15578500   0.910000  
..        ...        ...  
245  12963900   0.949998  
246  15549900   0.920000  
247  20946900   0.950001  
248  14139200   1.039999  
249  12490700   1.280000  

[250 rows x 8 columns]

In [ ]:
df.eval('High - Low')

0      0.990000
1      0.469999
2      0.590000
3      0.580000
4      0.910000
         ...   
245    0.949998
246    0.920000
247    0.950001
248    1.039999
249    1.280000
Length: 250, dtype: float64

In [ ]:
df.select_dtypes(exclude=object).pct_change()

Open      High       Low     Close  Adj Close    Volume
0         NaN       NaN       NaN       NaN        NaN       NaN
1    0.026938  0.007685  0.025132  0.000644   0.000644 -0.547909
2   -0.007997 -0.004131 -0.008065 -0.003541  -0.003541  0.418786
3   -0.001935 -0.006701 -0.006504 -0.008078  -0.008078  1.980294
4   -0.003554  0.017347  0.006874  0.026710   0.026710 -0.543915
..        ...       ...       ...       ...        ...       ...
245  0.001609  0.023553  0.015214  0.022818   0.022818 -0.539038
246  0.025301  0.027301  0.029567  0.024266   0.024266  0.199477
247  0.040345  0.019742  0.019276  0.026366   0.026366  0.347076
248  0.020331  0.032018  0.029718  0.011541   0.011541 -0.324998
249  0.010701  0.031746  0.023988  0.051159   0.051159 -0.116591

[250 rows x 6 columns]

In [ ]:
df.prod(numeric_only=True)

/usr/local/lib/python3.8/dist-packages/numpy/core/_methods.py:52: RuntimeWarning: overflow encountered in reduce
  return umr_prod(a, axis, dtype, out, keepdims, initial, where)


Open         inf
High         inf
Low          inf
Close        inf
Adj Close    inf
Volume       0.0
dtype: float64

In [ ]:
df.quantile([0, 0.1, 0.25, 0.5, 0.75, 0.9, 1], numeric_only=True)

Open       High        Low      Close  Adj Close      Volume
0.00  23.799999  25.049999  23.610001  24.290001  18.747433   4878800.0
0.10  29.509000  30.385999  29.014000  29.684001  21.131018  10529650.0
0.25  31.469999  31.910000  30.959999  31.527500  22.016315  13424175.0
0.50  34.370001  34.970000  33.784999  34.430000  24.288059  17982650.0
0.75  36.107499  36.797500  35.355001  36.235002  29.050100  24479175.0
0.90  37.400002  37.811001  36.600998  37.220001  32.311985  30529430.0
1.00  40.430000  42.080002  40.130001  41.560001  36.963558  69775500.0

In [ ]:
df.rank()

Date  Open  High   Low  Close  Adj Close  Volume
0      1.0  37.0  45.0  44.0   55.0        7.0   125.0
1      2.0  60.0  55.0  65.0   56.0        8.0    11.0
2      3.0  53.0  49.0  60.0   53.0        6.0    37.0
3      4.0  51.0  41.0  52.0   46.0        3.0   235.0
4      5.0  48.0  58.0  61.0   63.0        9.0    97.0
..     ...   ...   ...   ...    ...        ...     ...
245  246.0   3.0   3.0   4.0    4.0      148.0    58.0
246  247.0   4.0   4.0   5.0    5.0      153.0    96.0
247  248.0   6.0   6.0   7.0    6.0      158.0   161.0
248  249.0   9.5  10.0  10.0   10.0      166.0    75.0
249  250.0  12.0  15.0  15.0   17.0      182.0    55.0

[250 rows x 7 columns]

In [ ]:
df.round(3)

Date   Open   High    Low  Close  Adj Close    Volume
0    2021-12-27  30.44  31.23  30.24  31.04     19.029  17868100
1    2021-12-28  31.26  31.47  31.00  31.06     19.042   8078000
2    2021-12-29  31.01  31.34  30.75  30.95     18.974  11460950
3    2021-12-30  30.95  31.13  30.55  30.70     18.821  34157000
4    2022-01-03  30.84  31.67  30.76  31.52     19.324  15578500
..          ...    ...    ...    ...    ...        ...       ...
245  2022-12-19  24.90  25.64  24.69  25.55     25.550  12963900
246  2022-12-20  25.53  26.34  25.42  26.17     26.170  15549900
247  2022-12-21  26.56  26.86  25.91  26.86     26.860  20946900
248  2022-12-22  27.10  27.72  26.68  27.17     27.170  14139200
249  2022-12-23  27.39  28.60  27.32  28.56     28.560  12490700

[250 rows x 7 columns]

In [ ]:
df.var()

<ipython-input-304-28ded241fd7c>:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df.var()


Open         9.998785e+00
High         9.988790e+00
Low          9.731999e+00
Close        9.889970e+00
Adj Close    1.866652e+01
Volume       8.416312e+13
dtype: float64

In [ ]:
df.nunique()

Date         250
Open         217
High         228
Low          226
Close        222
Adj Close    242
Volume       250
dtype: int64

In [ ]:
df.High.value_counts()

35.000000    4
35.150002    4
37.730000    2
34.619999    2
37.459999    2
            ..
36.860001    1
36.750000    1
37.189999    1
37.880001    1
28.600000    1
Name: High, Length: 228, dtype: int64

In [ ]:
df.Volume.value_counts()

17868100    1
36015200    1
52964400    1
29426900    1
12902900    1
           ..
24557500    1
17397400    1
25131000    1
17456300    1
12490700    1
Name: Volume, Length: 250, dtype: int64

In [ ]:
df.Open.value_counts()

36.000000    4
32.910000    3
33.110001    3
34.000000    3
34.810001    3
            ..
36.520000    1
36.720001    1
36.740002    1
38.000000    1
27.389999    1
Name: Open, Length: 217, dtype: int64

* ### Seleção e indexação dos dados

Podemos selecionar de diversas formas um subconjuto da base de dados. Em boa parte dos casos vamos querer selecionar apenas uma coluna.

Assim temos para a coluna "High":

In [ ]:
df['High'].head()

0    31.230000
1    31.469999
2    31.340000
3    31.129999
4    31.670000
Name: High, dtype: float64

Podemos alterar algumas configurações do Pandas, por exemplo, na parte de visualização. Utilizamos a função `pd.set_option` para isto e vamos configurar para que mostre todas as linhas do dataframe a partir do parâmetro `display.max_rows`. 

Parar voltar a configuração padrão utilizamos a função `pd.reset_option`. E para saber quais são as configurações atuais utilizamos `pd.get_option`.

In [ ]:
pd.set_option("display.max_rows", None) # ou pd.set_option(max_rows, None)
df['High']

0      31.230000
1      31.469999
2      31.340000
3      31.129999
4      31.670000
5      32.189999
6      32.060001
7      31.430000
8      30.990000
9      31.059999
10     32.090000
11     33.250000
12     34.450001
13     34.740002
14     34.939999
15     34.930000
16     35.150002
17     34.840000
18     34.820000
19     35.080002
20     36.020000
21     37.400002
22     37.660000
23     37.799999
24     36.169998
25     36.459999
26     36.619999
27     35.480000
28     36.580002
29     35.950001
30     35.299999
31     35.900002
32     36.099998
33     37.259998
34     37.389999
35     36.049999
36     37.060001
37     36.580002
38     36.180000
39     37.080002
40     37.639999
41     36.799999
42     37.650002
43     36.480000
44     38.080002
45     37.889999
46     37.599998
47     37.500000
48     35.540001
49     35.119999
50     36.560001
51     36.130001
52     35.150002
53     34.009998
54     34.090000
55     34.320000
56     33.549999
57     34.540001
58     34.6199

In [ ]:
pd.reset_option("display.max_rows")

In [ ]:
df[['Open', 'Close']]

Open      Close
0    30.440001  31.040001
1    31.260000  31.059999
2    31.010000  30.950001
3    30.950001  30.700001
4    30.840000  31.520000
..         ...        ...
245  24.900000  25.549999
246  25.530001  26.170000
247  26.559999  26.860001
248  27.100000  27.170000
249  27.389999  28.559999

[250 rows x 2 columns]

Funções que usamos para selecionar subconjuntos dos dados:

* `df.iloc()`: retorna os elementos ou as linhas especificadas pelo argumento (numérico)
* `df.loc()`: retorna os elementos conforme labels especificadas no argumento (string) 
* `df.isin()`: verifica se os valores do datefram está contido no argumento
* `df.where(condition, other)`: substitui os valores onde a condição é Falsa 
* `df.mask()`: é o contrário do de cima, substitui os valores onde a condição é Verdadeira
* `df.insert()`: insere uma coluna nova no dataframe
* `df.pop()`: mostra uma coluna e exclui do dataframe

In [ ]:
# retorna a linha i (4) como series
df.iloc[4]

Date         2022-01-03
Open              30.84
High              31.67
Low               30.76
Close             31.52
Adj Close     19.323711
Volume         15578500
Name: 4, dtype: object

In [ ]:
# retorna a linha i como dataframe
df.iloc[[4]]

Date   Open   High    Low  Close  Adj Close    Volume
4  2022-01-03  30.84  31.67  30.76  31.52  19.323711  15578500

In [ ]:
# retorna o elemento i,j (4,0)
b = df.iloc[4,0]
b

'2022-01-03'

In [ ]:
df.iloc[4,0] = b

In [ ]:
# retorna as linhas 4 e 0 como dataframe
df.iloc[[4,0]]

Date       Open   High    Low      Close  Adj Close    Volume
4  2022-01-03  30.840000  31.67  30.76  31.520000  19.323711  15578500
0  2021-12-27  30.440001  31.23  30.24  31.040001  19.029440  17868100

In [ ]:
# retorna as linhas dos índices extremos
df.iloc[2:6]

Date       Open       High        Low      Close  Adj Close    Volume
2  2021-12-29  31.010000  31.340000  30.750000  30.950001  18.974266  11460950
3  2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001  34157000
4  2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711  15578500
5  2022-01-04  31.570000  32.189999  31.280001  31.920000  19.568937  17337100

In [ ]:
# retorna as linhas de 1 a 3 e as colunas de 0 a 2
df.iloc[1:3, 0:2]

Date   Open
1  2021-12-28  31.26
2  2021-12-29  31.01

In [ ]:
# retorna a linha 4 pois é o label da quarta linha
df.loc[4]

Date         2022-01-03
Open              30.84
High              31.67
Low               30.76
Close             31.52
Adj Close     19.323711
Volume         15578500
Name: 4, dtype: object

In [ ]:
a = df.iloc[0,0]
a

'2021-12-27'

In [ ]:
# mostra que o valor da primeira célula só aparece na primeira célula
df.isin([a])

Date   Open   High    Low  Close  Adj Close  Volume
0     True  False  False  False  False      False   False
1    False  False  False  False  False      False   False
2    False  False  False  False  False      False   False
3    False  False  False  False  False      False   False
4    False  False  False  False  False      False   False
..     ...    ...    ...    ...    ...        ...     ...
245  False  False  False  False  False      False   False
246  False  False  False  False  False      False   False
247  False  False  False  False  False      False   False
248  False  False  False  False  False      False   False
249  False  False  False  False  False      False   False

[250 rows x 7 columns]

In [ ]:
# o sinal '~' define negação
~df.isin([a])

Date  Open  High   Low  Close  Adj Close  Volume
0    False  True  True  True   True       True    True
1     True  True  True  True   True       True    True
2     True  True  True  True   True       True    True
3     True  True  True  True   True       True    True
4     True  True  True  True   True       True    True
..     ...   ...   ...   ...    ...        ...     ...
245   True  True  True  True   True       True    True
246   True  True  True  True   True       True    True
247   True  True  True  True   True       True    True
248   True  True  True  True   True       True    True
249   True  True  True  True   True       True    True

[250 rows x 7 columns]

In [ ]:
lista = []
for i in range(5):
  lista.append(df.iloc[i, 0])
df.isin(lista)

Date   Open   High    Low  Close  Adj Close  Volume
0     True  False  False  False  False      False   False
1     True  False  False  False  False      False   False
2     True  False  False  False  False      False   False
3     True  False  False  False  False      False   False
4     True  False  False  False  False      False   False
..     ...    ...    ...    ...    ...        ...     ...
245  False  False  False  False  False      False   False
246  False  False  False  False  False      False   False
247  False  False  False  False  False      False   False
248  False  False  False  False  False      False   False
249  False  False  False  False  False      False   False

[250 rows x 7 columns]

In [ ]:
m = df.High > 29.5
df_high = df.High
df_high.where(m, -df_high)

0      31.230000
1      31.469999
2      31.340000
3      31.129999
4      31.670000
         ...    
245   -25.639999
246   -26.340000
247   -26.860001
248   -27.719999
249   -28.600000
Name: High, Length: 250, dtype: float64

In [ ]:
df_high.mask(m, -df_high)

0     -31.230000
1     -31.469999
2     -31.340000
3     -31.129999
4     -31.670000
         ...    
245    25.639999
246    26.340000
247    26.860001
248    27.719999
249    28.600000
Name: High, Length: 250, dtype: float64

In [ ]:
Amplitude = df.High - df.Low
Amplitude

0      0.990000
1      0.469999
2      0.590000
3      0.580000
4      0.910000
         ...   
245    0.949998
246    0.920000
247    0.950001
248    1.039999
249    1.280000
Length: 250, dtype: float64

In [ ]:
df.insert(4, 'Amplitude', Amplitude)
df.head()

Date       Open       High        Low  Amplitude      Close  \
0  2021-12-27  30.440001  31.230000  30.240000   0.990000  31.040001   
1  2021-12-28  31.260000  31.469999  31.000000   0.469999  31.059999   
2  2021-12-29  31.010000  31.340000  30.750000   0.590000  30.950001   
3  2021-12-30  30.950001  31.129999  30.549999   0.580000  30.700001   
4  2022-01-03  30.840000  31.670000  30.760000   0.910000  31.520000   

   Adj Close    Volume  
0  19.029440  17868100  
1  19.041702   8078000  
2  18.974266  11460950  
3  18.821001  34157000  
4  19.323711  15578500

In [ ]:
# uma outra forma de definir uma nova coluna
df['Ampltude de fechamento'] = df.Open - df.Close
df.head()

Date       Open       High        Low  Amplitude      Close  \
0  2021-12-27  30.440001  31.230000  30.240000   0.990000  31.040001   
1  2021-12-28  31.260000  31.469999  31.000000   0.469999  31.059999   
2  2021-12-29  31.010000  31.340000  30.750000   0.590000  30.950001   
3  2021-12-30  30.950001  31.129999  30.549999   0.580000  30.700001   
4  2022-01-03  30.840000  31.670000  30.760000   0.910000  31.520000   

   Adj Close    Volume  Ampltude de fechamento  
0  19.029440  17868100               -0.600000  
1  19.041702   8078000                0.200001  
2  18.974266  11460950                0.059999  
3  18.821001  34157000                0.250000  
4  19.323711  15578500               -0.680000

In [ ]:
df.pop('Ampltude de fechamento')

0     -0.600000
1      0.200001
2      0.059999
3      0.250000
4     -0.680000
         ...   
245   -0.649999
246   -0.639999
247   -0.300002
248   -0.070000
249   -1.170000
Name: Ampltude de fechamento, Length: 250, dtype: float64

In [ ]:
df.pop('Amplitude')

0      0.990000
1      0.469999
2      0.590000
3      0.580000
4      0.910000
         ...   
245    0.949998
246    0.920000
247    0.950001
248    1.039999
249    1.280000
Name: Amplitude, Length: 250, dtype: float64

In [ ]:
df.head()

Date       Open       High        Low      Close  Adj Close    Volume
0  2021-12-27  30.440001  31.230000  30.240000  31.040001  19.029440  17868100
1  2021-12-28  31.260000  31.469999  31.000000  31.059999  19.041702   8078000
2  2021-12-29  31.010000  31.340000  30.750000  30.950001  18.974266  11460950
3  2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001  34157000
4  2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711  15578500

* ### Filtragem, ordenção e groupby

**Filtragem:** Podemos filtrar valores da seguinte forma por exemplo `df[df['Open'] > 30.32]` ou podemos usar operadores booleanos, como por exemplo `df.[df.Open < 31 & df.High > 31 | df.Close < 30]]`. A condição dentro do colchetes gera uma serie booleana e após a aplicação obtemos um dataframe apenas com as linhas com valor verdadeiro.

**Ordenação:** 
Vamos utilizar as funções:

* `df.sort_values(col, ascending = True)`: ordena os valores com base na coluna especificada 
* `df.nlargest(n, columns)`: mostra um dataframe com os n maiores valores da coluna especificada
* `df.nsmallest(n, columns)`: semelhante ao de cima, porém com os menores valores
* `df.transpose()`: transposta o dataframe
* `df.pivot()`: altera a estrutura do dateframe, "girando-o"
* `df.melt()`: faz o contrário da função acima


**Groupby:**
Aqui vamos utilizar a própria função groupby e também podemos apresentar a função apply:

* `df.groupby()`: divide os dados em grupos com base em alguma condição
* `df.apply()`: aplica uma função ao dataframe ou uma parte dele, podendo ser uma função de algum pacote ou função criada no próprio ambiente.


In [ ]:
df['Open'] > 30.32

0       True
1       True
2       True
3       True
4       True
       ...  
245    False
246    False
247    False
248    False
249    False
Name: Open, Length: 250, dtype: bool

In [ ]:
df[df['Open'] > 30.32]

Date       Open       High        Low      Close  Adj Close  \
0    2021-12-27  30.440001  31.230000  30.240000  31.040001  19.029440   
1    2021-12-28  31.260000  31.469999  31.000000  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000  30.750000  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
221  2022-11-14  31.070000  32.020000  30.959999  31.549999  28.060640   
222  2022-11-16  31.750000  31.980000  30.879999  31.200001  27.749352   
223  2022-11-17  31.200001  31.290001  30.469999  30.959999  27.535894   
224  2022-11-18  31.160000  31.389999  30.040001  30.389999  27.028934   
225  2022-11-21  30.940001  31.290001  29.530001  30.280001  26.931101   

       Volume  
0    17868100  
1     8078000  
2    11460950  
3    34157000  
4    15578500  
..        ...  
221  19816400  
222  25165000  
223  30514800  
224  27497600  
225  33579600  

[217 rows x 7 columns]

In [ ]:
df[(df.Open < 31) & (df.High > 31) | (df.Close < 30)]

Date       Open       High        Low      Close  Adj Close  \
0    2021-12-27  30.440001  31.230000  30.240000  31.040001  19.029440   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711   
7    2022-01-06  30.740000  31.430000  30.459999  30.580000  18.747433   
9    2022-01-10  30.700001  31.059999  30.389999  30.719999  18.833260   
10   2022-01-11  30.850000  32.090000  30.650000  31.990000  19.611851   
118  2022-06-17  31.100000  31.410000  28.840000  29.930000  22.020102   
120  2022-06-21  30.650000  30.770000  29.639999  29.870001  21.975958   
121  2022-06-22  29.360001  30.320000  29.040001  29.730000  21.872957   
122  2022-06-23  29.770000  30.200001  28.799999  29.100000  21.409454   
123  2022-06-24  29.510000  29.670000  28.830000  28.910000  21.269669   
124  2022-06-27  29.200001  31.040001  29.200001  30.860001  22.704321   
128  2022-07-01  30.600000  31.410000  30.000000  31.110001  22.888252   
132  2022-07-07  30.680000  31.209999  30.610001  30.920000  22.748465   
134  2022-07-11  30.780001  31.299999  30.620001  31.180000  22.939751   
136  2022-07-13  30.400000  31.010000  30.320000  30.700001  22.586607   
137  2022-07-14  30.320000  30.350000  29.570000  29.719999  21.865599   
139  2022-07-18  30.719999  31.400000  30.719999  31.320000  23.042751   
218  2022-11-09  30.520000  31.180000  30.000000  30.059999  26.735432   
219  2022-11-10  29.309999  30.240000  28.920000  29.690001  26.406353   
220  2022-11-11  29.610001  31.160000  29.280001  30.700001  27.304651   
225  2022-11-21  30.940001  31.290001  29.530001  30.280001  26.931101   
226  2022-11-22  26.719999  27.059999  25.530001  27.059999  27.059999   
227  2022-11-23  26.610001  27.410000  26.480000  27.070000  27.070000   
228  2022-11-24  27.100000  28.410000  26.940001  28.120001  28.120001   
229  2022-11-25  28.080000  28.080000  27.219999  27.490000  27.490000   
230  2022-11-28  27.379999  28.370001  27.219999  28.170000  28.170000   
231  2022-11-29  28.400000  29.959999  28.260000  29.379999  29.379999   
233  2022-12-01  30.219999  30.549999  29.250000  29.250000  29.250000   
234  2022-12-02  29.240000  30.450001  29.020000  29.629999  29.629999   
235  2022-12-05  29.700001  30.389999  29.170000  29.230000  29.230000   
236  2022-12-06  29.500000  30.000000  28.959999  29.180000  29.180000   
237  2022-12-07  29.170000  29.590000  28.719999  28.740000  28.740000   
238  2022-12-08  28.740000  29.309999  27.950001  28.120001  28.120001   
239  2022-12-09  28.240000  28.709999  27.799999  28.059999  28.059999   
240  2022-12-12  28.000000  28.080000  26.610001  27.299999  27.299999   
241  2022-12-13  27.459999  27.709999  26.879999  26.930000  26.930000   
242  2022-12-14  26.230000  26.360001  23.610001  24.290001  24.290001   
243  2022-12-15  23.799999  25.440001  23.709999  24.910000  24.910000   
244  2022-12-16  24.860001  25.049999  24.320000  24.980000  24.980000   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001  25.910000  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999  26.680000  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000  27.320000  28.559999  28.559999   

       Volume  
0    17868100  
3    34157000  
4    15578500  
7    30193600  
9    17334600  
10   30661100  
118  37348400  
120  23224300  
121  13553800  
122  14734000  
123  15527600  
124  17707400  
128  13968700  
132  10338500  
134  10408500  
136   8369200  
137  11563000  
139  13368200  
218  15928500  
219  39818600  
220  34866200  
225  33579600  
226  44857900  
227  17851100  
228  13776300  
229  13019700  
230  12320700  
231  24509100  
233  12067600  
234  18444400  
235  12452000  
236  10532200  
237  13339200  
238  13407800  
239  10363900  
240  18581300  
241  22207900 

In [ ]:
df.sort_values('High', ascending = False)

Date       Open       High        Low      Close  Adj Close  \
206  2022-10-21  40.430000  42.080002  40.130001  41.560001  36.963558   
207  2022-10-24  40.349998  40.619999  37.360001  37.450001  33.308117   
157  2022-08-11  40.180000  40.509998  38.860001  38.959999  28.663654   
205  2022-10-20  39.660000  40.290001  39.430000  40.189999  35.745075   
156  2022-08-10  40.250000  40.250000  39.270000  39.709999  29.215445   
..          ...        ...        ...        ...        ...        ...   
242  2022-12-14  26.230000  26.360001  23.610001  24.290001  24.290001   
246  2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
243  2022-12-15  23.799999  25.440001  23.709999  24.910000  24.910000   
244  2022-12-16  24.860001  25.049999  24.320000  24.980000  24.980000   

       Volume  
206  32623200  
207  33545600  
157  31853400  
205  27337800  
156  15755100  
..        ...  
242  69775500  
246  15549900  
245  12963900  
243  24269200  
244  28123600  

[250 rows x 7 columns]

In [ ]:
df.sort_values('Close', ascending = True)

Date       Open       High        Low      Close  Adj Close  \
242  2022-12-14  26.230000  26.360001  23.610001  24.290001  24.290001   
243  2022-12-15  23.799999  25.440001  23.709999  24.910000  24.910000   
244  2022-12-16  24.860001  25.049999  24.320000  24.980000  24.980000   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000   
..          ...        ...        ...        ...        ...        ...   
154  2022-08-08  37.619999  39.400002  37.619999  39.330002  28.935871   
156  2022-08-10  40.250000  40.250000  39.270000  39.709999  29.215445   
155  2022-08-09  39.740002  40.099998  39.439999  39.849998  29.318443   
205  2022-10-20  39.660000  40.290001  39.430000  40.189999  35.745075   
206  2022-10-21  40.430000  42.080002  40.130001  41.560001  36.963558   

       Volume  
242  69775500  
243  24269200  
244  28123600  
245  12963900  
246  15549900  
..        ...  
154  24523600  
156  15755100  
155  16911500  
205  27337800  
206  32623200  

[250 rows x 7 columns]

In [ ]:
df.nlargest(5, 'High')

Date       Open       High        Low      Close  Adj Close  \
206  2022-10-21  40.430000  42.080002  40.130001  41.560001  36.963558   
207  2022-10-24  40.349998  40.619999  37.360001  37.450001  33.308117   
157  2022-08-11  40.180000  40.509998  38.860001  38.959999  28.663654   
205  2022-10-20  39.660000  40.290001  39.430000  40.189999  35.745075   
156  2022-08-10  40.250000  40.250000  39.270000  39.709999  29.215445   

       Volume  
206  32623200  
207  33545600  
157  31853400  
205  27337800  
156  15755100

In [ ]:
df.nsmallest(5, 'High')

Date       Open       High        Low      Close  Adj Close  \
244  2022-12-16  24.860001  25.049999  24.320000  24.980000  24.980000   
243  2022-12-15  23.799999  25.440001  23.709999  24.910000  24.910000   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000   
242  2022-12-14  26.230000  26.360001  23.610001  24.290001  24.290001   

       Volume  
244  28123600  
243  24269200  
245  12963900  
246  15549900  
242  69775500

In [ ]:
df.transpose()

0           1           2           3           4    \
Date       2021-12-27  2021-12-28  2021-12-29  2021-12-30  2022-01-03   
Open        30.440001       31.26       31.01   30.950001       30.84   
High            31.23   31.469999       31.34   31.129999       31.67   
Low             30.24        31.0       30.75   30.549999       30.76   
Close       31.040001   31.059999   30.950001   30.700001       31.52   
Adj Close    19.02944   19.041702   18.974266   18.821001   19.323711   
Volume       17868100     8078000    11460950    34157000    15578500   

                  5           6           7           8           9    ...  \
Date       2022-01-04  2022-01-05  2022-01-06  2022-01-07  2022-01-10  ...   
Open            31.57        32.0       30.74       30.76   30.700001  ...   
High        32.189999   32.060001       31.43       30.99   31.059999  ...   
Low         31.280001   30.559999   30.459999   30.530001   30.389999  ...   
Close           31.92   30.610001       30.58       30.83   30.719999  ...   
Adj Close   19.568937   18.765823   18.747433     18.9007    18.83326  ...   
Volume       17337100    29552100    30193600    16889100    17334600  ...   

                  240         241         242         243         244  \
Date       2022-12-12  2022-12-13  2022-12-14  2022-12-15  2022-12-16   
Open             28.0   27.459999       26.23   23.799999   24.860001   
High            28.08   27.709999   26.360001   25.440001   25.049999   
Low         26.610001   26.879999   23.610001   23.709999       24.32   
Close       27.299999       26.93   24.290001       24.91       24.98   
Adj Close   27.299999       26.93   24.290001       24.91       24.98   
Volume       18581300    22207900    69775500    24269200    28123600   

                  245         246         247         248         249  
Date       2022-12-19  2022-12-20  2022-12-21  2022-12-22  2022-12-23  
Open             24.9   25.530001   26.559999        27.1   27.389999  
High        25.639999       26.34   26.860001   27.719999        28.6  
Low         24.690001       25.42       25.91       26.68       27.32  
Close       25.549999       26.17   26.860001       27.17   28.559999  
Adj Close   25.549999       26.17   26.860001       27.17   28.559999  
Volume       12963900    15549900    20946900    14139200    12490700  

[7 rows x 250 columns]

In [ ]:
produtos = pd.DataFrame({'Categoria': ['Limpeza', 'Limpeza', 'Entretenimento', 'Entretenimento', 'Tech', 'Tech'], 
                         'Preço': [10, 15, 20, 25, 50, 80], 
                         'Loja': ['Walmart', 'Dia', 'Walmart', 'Santo', 'Dia', 'Walmart'],
                         'Score': [8, 7, 9, 8, 9, 9]})
produtos

Categoria  Preço     Loja  Score
0         Limpeza     10  Walmart      8
1         Limpeza     15      Dia      7
2  Entretenimento     20  Walmart      9
3  Entretenimento     25    Santo      8
4            Tech     50      Dia      9
5            Tech     80  Walmart      9

In [ ]:
produtos_pivo = produtos.pivot(index = 'Categoria', columns = 'Loja', values = 'Preço')
produtos_pivo

Loja             Dia  Santo  Walmart
Categoria                           
Entretenimento   NaN   25.0     20.0
Limpeza         15.0    NaN     10.0
Tech            50.0    NaN     80.0

In [ ]:
produtos_pivo.melt()

Loja  value
0      Dia    NaN
1      Dia   15.0
2      Dia   50.0
3    Santo   25.0
4    Santo    NaN
5    Santo    NaN
6  Walmart   20.0
7  Walmart   10.0
8  Walmart   80.0

In [ ]:
df.groupby(['Date']).mean()

Open       High        Low      Close  Adj Close      Volume
Date                                                                         
2021-12-27  30.440001  31.230000  30.240000  31.040001  19.029440  17868100.0
2021-12-28  31.260000  31.469999  31.000000  31.059999  19.041702   8078000.0
2021-12-29  31.010000  31.340000  30.750000  30.950001  18.974266  11460950.0
2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001  34157000.0
2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711  15578500.0
...               ...        ...        ...        ...        ...         ...
2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999  12963900.0
2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000  15549900.0
2022-12-21  26.559999  26.860001  25.910000  26.860001  26.860001  20946900.0
2022-12-22  27.100000  27.719999  26.680000  27.170000  27.170000  14139200.0
2022-12-23  27.389999  28.600000  27.320000  28.559999  28.559999  12490700.0

[250 rows x 6 columns]

Aqui calculamos a média das outras colunas para cada data. Porém como há 250 datas diferentes o data frame possui 250 linhas. Essa função é mais utilizado quando há features categóricas, onde queremos alguma informação de interesse de cada categoria.

In [ ]:
df.Volume.apply(np.sqrt)

0      4227.067541
1      2842.182260
2      3385.402487
3      5844.399028
4      3946.960856
          ...     
245    3600.541626
246    3943.336151
247    4576.778343
248    3760.212760
249    3534.218443
Name: Volume, Length: 250, dtype: float64

In [ ]:
df.Volume.apply(lambda x: x**(1/2))

0      4227.067541
1      2842.182260
2      3385.402487
3      5844.399028
4      3946.960856
          ...     
245    3600.541626
246    3943.336151
247    4576.778343
248    3760.212760
249    3534.218443
Name: Volume, Length: 250, dtype: float64

In [ ]:
def raiz(x):
  return x ** (1/2)
df.Volume.apply(lambda x: raiz(x))

0      4227.067541
1      2842.182260
2      3385.402487
3      5844.399028
4      3946.960856
          ...     
245    3600.541626
246    3943.336151
247    4576.778343
248    3760.212760
249    3534.218443
Name: Volume, Length: 250, dtype: float64

In [ ]:
df.Volume.apply(raiz)

0      4227.067541
1      2842.182260
2      3385.402487
3      5844.399028
4      3946.960856
          ...     
245    3600.541626
246    3943.336151
247    4576.778343
248    3760.212760
249    3534.218443
Name: Volume, Length: 250, dtype: float64

* ### Manipulação de dados ausentes

Algo que é bem comum nas bases de dados é a ausência de valores. Posto isto, é necessário ter algum tratamento com esses dados, seja: removendo as observações com dados faltantes, eliminando do modelo as features que possuam muitos valores ausentes, preenchendo com a média (ou outro valor) todos os dados que faltam etc.

E claro, o Pandas possui diversas funções que são úteis para esse tratamento, são elas:

* `df.isnull()`: retorna um dataframe com valores booleanos, sendo verdadeiro  no caso de haver valores ausentes 
* `df.notnull()`: retorna o inverso da função acima 
* `df.dropna(axis = index)`: remove as linhas (por padrão) que possuma valores faltantes
* `df.fillna()`: preenche os valores ausentes com o valor passado na função
* `df.interpoalte()`: preenche os valores ausentes com uma interpolação que depende do método passado
* `df.replace()`: substitui os valores passados no argumento, por exemplo, substituir os valores booleanos Falses/True por 0/1 ou gênero masculino e feminino por 0/1


In [ ]:
df.isnull()

Date   Open   High    Low  Close  Adj Close  Volume
0    False  False  False  False  False      False   False
1    False  False  False  False  False      False   False
2    False  False  False  False  False      False   False
3    False  False  False  False  False      False   False
4    False  False  False  False  False      False   False
..     ...    ...    ...    ...    ...        ...     ...
245  False  False  False  False  False      False   False
246  False  False  False  False  False      False   False
247  False  False  False  False  False      False   False
248  False  False  False  False  False      False   False
249  False  False  False  False  False      False   False

[250 rows x 7 columns]

In [ ]:
df.isnull().sum()

Date         0
Open         0
High         0
Low          0
Close        0
Adj Close    0
Volume       0
dtype: int64

In [ ]:
df.notnull().sum()

Date         250
Open         250
High         250
Low          250
Close        250
Adj Close    250
Volume       250
dtype: int64

In [ ]:
df2 = df.copy()
df2.iloc[0,0] = pd.NA
df2.dropna()

Date       Open       High        Low      Close  Adj Close  \
1    2021-12-28  31.260000  31.469999  31.000000  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000  30.750000  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711   
5    2022-01-04  31.570000  32.189999  31.280001  31.920000  19.568937   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001  25.910000  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999  26.680000  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000  27.320000  28.559999  28.559999   

       Volume  
1     8078000  
2    11460950  
3    34157000  
4    15578500  
5    17337100  
..        ...  
245  12963900  
246  15549900  
247  20946900  
248  14139200  
249  12490700  

[249 rows x 7 columns]

In [ ]:
df2.iloc[1,3] = pd.NA
df2.dropna(axis='columns')

Open       High      Close  Adj Close    Volume
0    30.440001  31.230000  31.040001  19.029440  17868100
1    31.260000  31.469999  31.059999  19.041702   8078000
2    31.010000  31.340000  30.950001  18.974266  11460950
3    30.950001  31.129999  30.700001  18.821001  34157000
4    30.840000  31.670000  31.520000  19.323711  15578500
..         ...        ...        ...        ...       ...
245  24.900000  25.639999  25.549999  25.549999  12963900
246  25.530001  26.340000  26.170000  26.170000  15549900
247  26.559999  26.860001  26.860001  26.860001  20946900
248  27.100000  27.719999  27.170000  27.170000  14139200
249  27.389999  28.600000  28.559999  28.559999  12490700

[250 rows x 5 columns]

In [ ]:
df2.Low.fillna(df2.Low.mean())

0      30.240000
1      33.185703
2      30.750000
3      30.549999
4      30.760000
         ...    
245    24.690001
246    25.420000
247    25.910000
248    26.680000
249    27.320000
Name: Low, Length: 250, dtype: float64

In [ ]:
# ffill = forward fill preenche com os valores da célula anterior 
df2.Date.fillna(method='ffill')

0            <NA>
1      2021-12-28
2      2021-12-29
3      2021-12-30
4      2022-01-03
          ...    
245    2022-12-19
246    2022-12-20
247    2022-12-21
248    2022-12-22
249    2022-12-23
Name: Date, Length: 250, dtype: object

In [ ]:
# bfill = backward fill preenche com os valores da célula posterior 
df2.Date.fillna(method='bfill')

0      2021-12-28
1      2021-12-28
2      2021-12-29
3      2021-12-30
4      2022-01-03
          ...    
245    2022-12-19
246    2022-12-20
247    2022-12-21
248    2022-12-22
249    2022-12-23
Name: Date, Length: 250, dtype: object

In [ ]:
df2

Date       Open       High        Low      Close  Adj Close  \
0          <NA>  30.440001  31.230000      30.24  31.040001  19.029440   
1    2021-12-28  31.260000  31.469999       <NA>  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000      30.75  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000      30.76  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000      25.42  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001      25.91  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999      26.68  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000      27.32  28.559999  28.559999   

       Volume  
0    17868100  
1     8078000  
2    11460950  
3    34157000  
4    15578500  
..        ...  
245  12963900  
246  15549900  
247  20946900  
248  14139200  
249  12490700  

[250 rows x 7 columns]

In [ ]:
for i in range(10):
  if i % 2 == 1:
    df2.iloc[i, 6] = pd.NA
df2

Date       Open       High        Low      Close  Adj Close  \
0          <NA>  30.440001  31.230000      30.24  31.040001  19.029440   
1    2021-12-28  31.260000  31.469999       <NA>  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000      30.75  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000      30.76  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000      25.42  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001      25.91  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999      26.68  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000      27.32  28.559999  28.559999   

       Volume  
0    17868100  
1        <NA>  
2    11460950  
3        <NA>  
4    15578500  
..        ...  
245  12963900  
246  15549900  
247  20946900  
248  14139200  
249  12490700  

[250 rows x 7 columns]

In [ ]:
df2.Volume.interpolate(method='pad')

0      17868100
1      17868100
2      11460950
3      11460950
4      15578500
         ...   
245    12963900
246    15549900
247    20946900
248    14139200
249    12490700
Name: Volume, Length: 250, dtype: int64

In [ ]:
j = df2.iloc[1,1]
j

31.26

In [ ]:
df2.replace(j, 'AQUI')

Date       Open       High        Low      Close  Adj Close  \
0          <NA>  30.440001  31.230000      30.24  31.040001  19.029440   
1    2021-12-28       AQUI  31.469999       <NA>  31.059999  19.041702   
2    2021-12-29      31.01  31.340000      30.75  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03      30.84  31.670000      30.76  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19       24.9  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000      25.42  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001      25.91  26.860001  26.860001   
248  2022-12-22       27.1  27.719999      26.68  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000      27.32  28.559999  28.559999   

       Volume  
0    17868100  
1        <NA>  
2    11460950  
3        <NA>  
4    15578500  
..        ...  
245  12963900  
246  15549900  
247  20946900  
248  14139200  
249  12490700  

[250 rows x 7 columns]

* ### Agregar e combinar dados

Outra competência muito importante no trabalho com dados é a de agregar e combinar dados de diferentes fontes e dataframes.

Algumas funções que o Pandas entrega para esse trabalho:

* `df1.append(df2)`: adiciona as observações de df2 em df1 (é necessário que tenham colunas idênticas)
* `df.concat([df1, df2], axis=1)`: adicionas as colunas de df2 em df1 (as linhas devem ser idênticas)
* `df.assign()`: cria novas colunas no dataframe 
* `df.compare()`: compara os dados de dois dataframes e retorna coluna as linhas onde há diferenças
* `df1.update(df2)`: atualiza os de df1 com os dados de df2 (ambos os dataframes devem possuir os mesmos índices e colunas)

In [ ]:
df2.append(df)

Date       Open       High        Low      Close  Adj Close  \
0          <NA>  30.440001  31.230000      30.24  31.040001  19.029440   
1    2021-12-28  31.260000  31.469999       <NA>  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000      30.75  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000      30.76  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000      25.42  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001      25.91  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999      26.68  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000      27.32  28.559999  28.559999   

       Volume  
0    17868100  
1        <NA>  
2    11460950  
3        <NA>  
4    15578500  
..        ...  
245  12963900  
246  15549900  
247  20946900  
248  14139200  
249  12490700  

[500 rows x 7 columns]

In [ ]:
# semelhante ao de cima
pd.concat([df, df2], axis = 0)

Date       Open       High        Low      Close  Adj Close  \
0    2021-12-27  30.440001  31.230000      30.24  31.040001  19.029440   
1    2021-12-28  31.260000  31.469999       31.0  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000      30.75  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000      30.76  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000      25.42  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001      25.91  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999      26.68  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000      27.32  28.559999  28.559999   

       Volume  
0    17868100  
1     8078000  
2    11460950  
3    34157000  
4    15578500  
..        ...  
245  12963900  
246  15549900  
247  20946900  
248  14139200  
249  12490700  

[500 rows x 7 columns]

In [ ]:
pd.concat([df, df2], axis = 1)

Date       Open       High        Low      Close  Adj Close  \
0    2021-12-27  30.440001  31.230000  30.240000  31.040001  19.029440   
1    2021-12-28  31.260000  31.469999  31.000000  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000  30.750000  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001  25.910000  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999  26.680000  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000  27.320000  28.559999  28.559999   

       Volume        Date       Open       High        Low      Close  \
0    17868100        <NA>  30.440001  31.230000      30.24  31.040001   
1     8078000  2021-12-28  31.260000  31.469999       <NA>  31.059999   
2    11460950  2021-12-29  31.010000  31.340000      30.75  30.950001   
3    34157000  2021-12-30  30.950001  31.129999  30.549999  30.700001   
4    15578500  2022-01-03  30.840000  31.670000      30.76  31.520000   
..        ...         ...        ...        ...        ...        ...   
245  12963900  2022-12-19  24.900000  25.639999  24.690001  25.549999   
246  15549900  2022-12-20  25.530001  26.340000      25.42  26.170000   
247  20946900  2022-12-21  26.559999  26.860001      25.91  26.860001   
248  14139200  2022-12-22  27.100000  27.719999      26.68  27.170000   
249  12490700  2022-12-23  27.389999  28.600000      27.32  28.559999   

     Adj Close    Volume  
0    19.029440  17868100  
1    19.041702      <NA>  
2    18.974266  11460950  
3    18.821001      <NA>  
4    19.323711  15578500  
..         ...       ...  
245  25.549999  12963900  
246  26.170000  15549900  
247  26.860001  20946900  
248  27.170000  14139200  
249  28.559999  12490700  

[250 rows x 14 columns]

In [ ]:
df.assign(Variacao = (df['Open'] - df['Close'] ) / (df['High'] - df['Low']))

Date       Open       High        Low      Close  Adj Close  \
0    2021-12-27  30.440001  31.230000  30.240000  31.040001  19.029440   
1    2021-12-28  31.260000  31.469999  31.000000  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000  30.750000  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000  30.760000  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000  25.420000  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001  25.910000  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999  26.680000  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000  27.320000  28.559999  28.559999   

       Volume  Variacao  
0    17868100 -0.606061  
1     8078000  0.425535  
2    11460950  0.101693  
3    34157000  0.431034  
4    15578500 -0.747253  
..        ...       ...  
245  12963900 -0.684211  
246  15549900 -0.695651  
247  20946900 -0.315791  
248  14139200 -0.067308  
249  12490700 -0.914063  

[250 rows x 8 columns]

In [ ]:
df.compare(df2)

Date              Low               Volume          
           self other       self      other      self     other
0    2021-12-27  <NA>  30.240000      30.24  17868100  17868100
1           NaN   NaN  31.000000       <NA>   8078000      <NA>
2           NaN   NaN  30.750000      30.75  11460950  11460950
3           NaN   NaN  30.549999  30.549999  34157000      <NA>
4           NaN   NaN  30.760000      30.76  15578500  15578500
..          ...   ...        ...        ...       ...       ...
245         NaN   NaN  24.690001  24.690001  12963900  12963900
246         NaN   NaN  25.420000      25.42  15549900  15549900
247         NaN   NaN  25.910000      25.91  20946900  20946900
248         NaN   NaN  26.680000      26.68  14139200  14139200
249         NaN   NaN  27.320000      27.32  12490700  12490700

[250 rows x 6 columns]

In [ ]:
df2.loc[249, 'Low'] == df.loc[249, 'Low']

True

In [ ]:
df2

Date       Open       High        Low      Close  Adj Close  \
0          <NA>  30.440001  31.230000      30.24  31.040001  19.029440   
1    2021-12-28  31.260000  31.469999       <NA>  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000      30.75  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000      30.76  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000      25.42  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001      25.91  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999      26.68  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000      27.32  28.559999  28.559999   

       Volume  
0    17868100  
1        <NA>  
2    11460950  
3        <NA>  
4    15578500  
..        ...  
245  12963900  
246  15549900  
247  20946900  
248  14139200  
249  12490700  

[250 rows x 7 columns]

In [ ]:
df2.update(df)
df2

Date       Open       High        Low      Close  Adj Close  \
0    2021-12-27  30.440001  31.230000      30.24  31.040001  19.029440   
1    2021-12-28  31.260000  31.469999       31.0  31.059999  19.041702   
2    2021-12-29  31.010000  31.340000      30.75  30.950001  18.974266   
3    2021-12-30  30.950001  31.129999  30.549999  30.700001  18.821001   
4    2022-01-03  30.840000  31.670000      30.76  31.520000  19.323711   
..          ...        ...        ...        ...        ...        ...   
245  2022-12-19  24.900000  25.639999  24.690001  25.549999  25.549999   
246  2022-12-20  25.530001  26.340000      25.42  26.170000  26.170000   
247  2022-12-21  26.559999  26.860001      25.91  26.860001  26.860001   
248  2022-12-22  27.100000  27.719999      26.68  27.170000  27.170000   
249  2022-12-23  27.389999  28.600000      27.32  28.559999  28.559999   

       Volume  
0    17868100  
1     8078000  
2    11460950  
3    34157000  
4    15578500  
..        ...  
245  12963900  
246  15549900  
247  20946900  
248  14139200  
249  12490700  

[250 rows x 7 columns]